File: java/src/main/registry/bucketRegistry.java
package main.registry;

import java.util.concurrent.ConcurrentHashMap;

import main.core.tokenBucket;
import main.time.systemTimeProvider;

public class bucketRegistry {
    // a thread safe map that return correct bucket as per the key

    private final int bucketCapcity;
    private final long refillInterval;
    private final systemTimeProvider timeProvider;

    private final ConcurrentHashMap<String, tokenBucket> buckets = new ConcurrentHashMap<>();

    // configs

    public bucketRegistry(int bucketCapcity, long refillInterval, systemTimeProvider timeProvider) {
        this.bucketCapcity = bucketCapcity;
        this.refillInterval = refillInterval;
        this.timeProvider = timeProvider;
    }

    // get and creation of bucket

    public tokenBucket getBucket(String key) {
        return buckets.computeIfAbsent(key,
                k -> new tokenBucket(bucketCapcity, refillInterval, timeProvider));
    }

    public void showBuckets() {
        buckets.forEach((key, bucket) -> System.out.println(key + "->" + bucket));
    }
}

File: typescript/src/registry/BucketRegistry.ts
import { TokenBucket } from "../core/TokenBucket";
import { TimeProvider } from "../time/TimeProvider";

/* Registry for managing buckets per key */

export class BucketRegistry {
  private readonly bucketCapacity: number;
  private readonly refillInterval: number;
  private readonly timeProvider: TimeProvider;
  private readonly BUCKET_TTL_MS = 5 * 60 * 1000;

  private buckets: Map<String, TokenBucket> = new Map();

  constructor(
    bucketCapacity: number,
    refillInterval: number,
    timeProvider: TimeProvider
  ) {
    this.bucketCapacity = bucketCapacity;
    this.refillInterval = refillInterval;
    this.timeProvider = timeProvider;
  }

  /* @return existing bucket for the key or create new bucket */

  public getBucket(key: String): TokenBucket {
    const existBucket = this.getBucket(key);

    if (existBucket) {
      return existBucket;
    }

    const newBucket = new TokenBucket(
      this.bucketCapacity,
      this.refillInterval,
      this.timeProvider
    );

    this.buckets.set(key, newBucket);
    return newBucket;
  }

  /* remove unused buckets that are not accessed within TTL */

  public cleanupUnusedBuckets(): void {
    const currentTime = this.timeProvider.now();

    for (const [key, bucket] of this.buckets.entries()) {
      if (currentTime - bucket.getLastAccessTime() > this.BUCKET_TTL_MS) {
        this.buckets.delete(key);
      }
    }
  }
}
